From 75568e028a853ec407ab96e9e66fe6dabfefee51 Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Wed, 9 Dec 2020 10:20:00 +0800
Subject: [PATCH] pfeng-hif: disable softirq before napi_alloc_frag

To fix below call trace:
[  313.053231] BUG: using smp_processor_id() in preemptible [00000000] code: systemd-udevd/2438
[  313.061538] caller is napi_alloc_frag+0x2c/0x54
[  313.066300] CPU: 0 PID: 2438 Comm: systemd-udevd Tainted: G        W  O      5.4.80-yocto-standard #1
[  313.075264] Hardware name: Freescale S32G274 (DT)
[  313.079950] Call trace:
[  313.082387]  dump_backtrace+0x0/0x184
[  313.086031]  show_stack+0x28/0x34
[  313.089330]  dump_stack+0xd0/0x130
[  313.092716]  debug_smp_processor_id+0x13c/0x140
[  313.097226]  napi_alloc_frag+0x2c/0x54
[  313.101030]  pfeng_hif_chnl_refill_rx_buffer+0x34/0x260 [pfeng]
[  313.106898]  pfeng_hif_chnl_fill_rx_buffers+0x34/0x64 [pfeng]
[  313.112626]  pfeng_hif_client_add+0x144/0x260 [pfeng]
[  313.117662]  pfeng_napi_if_create+0x238/0x25c [pfeng]
[  313.122696]  pfeng_drv_probe+0x164/0x210 [pfeng]
[  313.127297]  pfeng_s32g_probe.part.0+0x158/0x17c [pfeng]
[  313.132592]  pfeng_s32g_probe+0x2c/0x50 [pfeng]
[  313.137069]  platform_drv_probe+0x5c/0xb0
[  313.141061]  really_probe+0x2c8/0x4c0
[  313.144704]  driver_probe_device+0xa8/0x140
[  313.148871]  device_driver_attach+0xc8/0xd0
[  313.153045]  __driver_attach+0x100/0x180
[  313.156947]  bus_for_each_dev+0x84/0xe0
[  313.160765]  driver_attach+0x34/0x40
[  313.164322]  bus_add_driver+0x138/0x224
[  313.168143]  driver_register+0x7c/0x124
[  313.171960]  __platform_driver_register+0x58/0x64
[  313.176686]  pfeng_platform_driver_init+0x28/0x1000 [pfeng]
[  313.182206]  do_one_initcall+0x58/0x194
[  313.186026]  do_init_module+0x60/0x274
[  313.189756]  load_module+0x58c/0x690
[  313.193314]  __do_sys_finit_module+0xe8/0x11c
[  313.197654]  __arm64_sys_finit_module+0x2c/0x40
[  313.202168]  el0_svc_common.constprop.0+0x78/0x13c
[  313.206938]  el0_svc_handler+0x54/0xc0
[  313.210671]  el0_svc+0x8/0x208

It needs to disable softirq before napi_alloc_frag to keep safe to
fix above call trace.

Upstream-Status: Inappropriate [Upstream with latest version, has replaced these
series code with other API functions.]

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 sw/linux-pfeng/pfeng-hif.c | 2 ++
 1 file changed, 2 insertions(+)
 mode change 100644 => 100755 sw/linux-pfeng/pfeng-hif.c

diff --git a/sw/linux-pfeng/pfeng-hif.c b/sw/linux-pfeng/pfeng-hif.c
old mode 100644
new mode 100755
index becebca..1647ef5
--- a/sw/linux-pfeng/pfeng-hif.c
+++ b/sw/linux-pfeng/pfeng-hif.c
@@ -60,7 +60,9 @@ static struct sk_buff *pfeng_bman_build_skb(struct pfeng_rx_chnl_pool *pool)
 	struct sk_buff *skb;
 	u8 *buf;
 
+	local_bh_disable();
 	buf = napi_alloc_frag(truesize);
+	local_bh_enable();
 	if (!buf) {
 		dev_err(pool->dev, "chnl%d: No mem for skb\n", pool->id);
 		return NULL;
-- 
2.17.1

